lib/mongo/auth/scram256/conversation.rb

```diff
@@ -0,0 +1,509 @@
+          @rnonce ||= payload_data.match(RNONCE)[1]
+        end
+
+        # Gets the salt from the server response.
+        #
+        # @api private
+        #
+        # @since 2.6.0
+        def salt
+          @salt ||= payload_data.match(SALT)[1]
+        end
+
+        # Salted password algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-3
+        #
+        # @since 2.6.0
+        def salted_password
+          @salted_password ||= hi(user.sasl_prepped_hashed_password)
+        end
+
+        # Server key algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-3
+        #
+        # @since 2.6.0
+        def server_key
+          @server_key ||= hmac(salted_password, SERVER_KEY)
+        end
+
+        # Server signature algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-3
+        #
+        # @since 2.6.0
+        def server_signature
+          @server_signature ||= Base64.strict_encode64(hmac(server_key, auth_message))
+        end
+
+        # Stored key algorithm implementation.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-3
+        #
+        # @since 2.6.0
+        def stored_key(key)
+          h(key)
+        end
+
+        # Get the verifier token from the server response.
+        #
+        # @api private
+        #
+        # @since 2.6.0
+        def verifier
+          @verifier ||= payload_data.match(VERIFIER)[1]
+        end
+
+        # Get the without proof message.
+        #
+        # @api private
+        #
+        # @see http://tools.ietf.org/html/rfc5802#section-7
+        #
+        # @since 2.6.0
+        def without_proof
+          @without_proof ||= "c=biws,r=#{rnonce}"
+        end
+
+        # XOR operation for two strings.
+        #
+        # @api private
+        #
+        # @since 2.6.0
+        def xor(first, second)
+          first.bytes.zip(second.bytes).map{ |(a,b)| (a ^ b).chr }.join('')
+        end
+
+        def compare_digest(a, b)
+          check = a.bytesize ^ b.bytesize
+          a.bytes.zip(b.bytes){ |x, y| check |= x ^ y.to_i }
+          check == 0
+        end
+
+        def validate_final_message!(reply)
+          validate!(reply)
+          unless compare_digest(verifier, server_signature)
+            raise Error::InvalidSignature.new(verifier, server_signature)
+          end
+        end
+
+        def validate_first_message!(reply)
+          validate!(reply)
+          raise Error::InvalidNonce.new(nonce, rnonce) unless rnonce.start_with?(nonce)
+        end
+
+        def validate!(reply)
+          raise Unauthorized.new(user) unless reply.documents[0][Operation::Result::OK] == 1
+          @reply = reply
+        end
+
+        def digest
+          @digest ||= DIGEST
```

Comment:
I am quite certain that this is not correct. The definition of DIGEST in SCRAM isn't correct either but it is not used anywhere (def digest in SCRAM does not use DIGEST constant) hence is effectively harmless.

DIGEST as defined in SCRAM and copied by you to SCRAM256 is a frozen instance of a digest class. Except freeze on these instances doesn't seem to do anything - they can be mutated with `update` even when frozen. And since you *are* using DIGEST here, what this code does is it uses the same digest instance for all digests in the application (original SCRAM does not have this problem as its digest method does not use DIGEST constant).

I have this repaired in https://github.com/p-mongo/mongo-ruby-driver/commit/e1875021d44b9a35a0657076c3d1db2cc3c59226 as part of DRYing the code.